## ICP7 - Question1
#### Change the classifier in the given source code to
1. SVM and see how accuracy changes
2. Setthe tfidf vectorizer parameter to use bigram and see how the accuracy changes TfidfVectorizer(ngram_range=(1,2))
3. Settfidf vectorizer argumentto use stop_words='english'and see how accuracy changes

In [1]:
#importing Libraries
from sklearn.datasets import fetch_20newsgroups

from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn import metrics
from sklearn.pipeline import Pipeline
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC

In [2]:
#Reading the data
twenty_train = fetch_20newsgroups(subset='train', shuffle=True)

In [3]:
#Transorming the data
tfidf_Vect = TfidfVectorizer()
X_train_tfidf = tfidf_Vect.fit_transform(twenty_train.data)

In [33]:
#Printing the vocabulary
#print(tfidf_Vect.vocabulary_)

In [5]:
#Buliding MultinomialNB() Classifier
clf = MultinomialNB()
clf.fit(X_train_tfidf, twenty_train.target)

MultinomialNB()

In [6]:
#Evaluate
twenty_test = fetch_20newsgroups(subset='test', shuffle=True)
X_test_tfidf = tfidf_Vect.transform(twenty_test.data)

In [7]:
#predecting the data
predicted = clf.predict(X_test_tfidf)

In [8]:
#finding the accuracy 
score = metrics.accuracy_score(twenty_test.target, predicted)
print('Multinomial score : ', score)

Multinomial score :  0.7738980350504514


#### 1. Changing the classifier to SVM and finding the accuracy

In [29]:
from sklearn import svm

In [30]:
#Buliding the SVC Classifier
#svc = KNeighborsClassifier(n_neighbors=2)
SVM = svm.SVC(C=1.0, kernel='linear', degree=3, gamma='auto')
svc.fit(X_train_tfidf, twenty_train.target)

KNeighborsClassifier(n_neighbors=2)

In [32]:
#Finding the accuracy using SVC Classifier
acc_knn = round(svc.score(X_train_tfidf, twenty_train.target) * 100, 2)
print("Accuracy with SVM Classifier is :", acc_knn / 100)

Accuracy with SVM Classifier is : 0.9257


#### 2. Set the tfidf vectorizer parameter to use bigram and see how the accuracy changes TfidfVectorizer(ngram_range=(1,2))

In [12]:
#setting the tfidf vectorizer parameter to bigram
tfidf_Vect_bigram = TfidfVectorizer(ngram_range=(1, 2))

In [13]:
#Transorming the data
X_train_tfidf_bigram = tfidf_Vect_bigram.fit_transform(twenty_train.data)

In [14]:
#Buliding MultinomialNB() Classifier
clf_bigram = MultinomialNB()
clf_bigram.fit(X_train_tfidf_bigram, twenty_train.target)

MultinomialNB()

In [15]:
#Evaluate
twenty_test_bigram = fetch_20newsgroups(subset='test', shuffle=True)
X_test_tfidf_bigram = tfidf_Vect_bigram.transform(twenty_test.data)

In [16]:
#Predecting the data
predicted_bigram = clf_bigram.predict(X_test_tfidf_bigram)

In [17]:
#Finding the Accuracy using MultinomialNB Classifer
score_bigram = round(metrics.accuracy_score(twenty_test_bigram.target, predicted_bigram), 4)
print("Accuracy for tf-idf vectorizer (bigram) using MultinomialNB classifier is : ", score_bigram)

Accuracy for tf-idf vectorizer (bigram) using MultinomialNB classifier is :  0.7654


In [18]:
#Buliding the SVC Classifier
svc = KNeighborsClassifier(n_neighbors=2)
svc.fit(X_train_tfidf_bigram, twenty_train.target)

KNeighborsClassifier(n_neighbors=2)

In [19]:
#Finding the accuracy using SVC Classifier
acc_knn_bigram = round(svc.score(X_train_tfidf_bigram, twenty_train.target) * 100, 2)
print("Accuracy for tf-idf vectorizer (bigram) using SVC classifier is :", acc_knn_bigram / 100)

Accuracy for tf-idf vectorizer (bigram) using SVC classifier is : 0.9209999999999999


#### 3.Set tfidf vectorizer argument to use stop_words='english'and see how accuracy changes

In [20]:
#setting the tfidf vectorizer parameter to use stop_word = 'english'
tfidf_Vect_stopwords = TfidfVectorizer(stop_words='english')

In [21]:
#Transorming the data
X_train_tfidf_stopwords = tfidf_Vect_stopwords.fit_transform(twenty_train.data)

In [22]:
#Buliding MultinomialNB() Classifier
clf_stopwords = MultinomialNB()
clf_stopwords.fit(X_train_tfidf_stopwords, twenty_train.target)

MultinomialNB()

In [23]:
#Evaluate
twenty_test_stopwords = fetch_20newsgroups(subset='test', shuffle=True)
X_test_tfidf_stopwords = tfidf_Vect_stopwords.transform(twenty_test.data)

In [24]:
#Predecting the data
predicted_stopwords = clf_stopwords.predict(X_test_tfidf_stopwords)

In [25]:
#Finding the Accuracy using MultinomialNB Classifer
score_stopwords = round(metrics.accuracy_score(twenty_test_bigram.target, predicted_stopwords), 4)
print("Accuracy for tf-idf vectorizer (Stopwords) using MultinomialNB classifier is : ", score_stopwords)

Accuracy for tf-idf vectorizer (Stopwords) using MultinomialNB classifier is :  0.8169


In [26]:
#Buliding the SVC Classifier
svc = KNeighborsClassifier(n_neighbors=2)
svc.fit(X_train_tfidf_stopwords, twenty_train.target)

KNeighborsClassifier(n_neighbors=2)

In [27]:
#Finding the accuracy using SVC Classifier
acc_knn_stopwords = round(svc.score(X_train_tfidf_stopwords, twenty_train.target) * 100, 2)
print("Accuracy for tf-idf vectorizer (bstopwords) using SVC classifier is :", acc_knn_stopwords / 100)

Accuracy for tf-idf vectorizer (bstopwords) using SVC classifier is : 0.9244


In [28]:
print('Accuracy using MultinomialNB Classifier : ' ,score)
print('Accuracy when tf-idf vectorizer paramater set to bigram using MultinominalNB Classifier: ' ,score_bigram)
print('Accuracy when tf-idf vectorizer paramater set to stopwords(English) using MultinominalNB Classifier: ' ,score_stopwords)
print('  ')
print('Accuracy using SVM Classifier : ' ,acc_knn / 100)
print('Accuracy when tf-idf vectorizer paramater set to bigram using SVM Classifier  : ' ,acc_knn_bigram / 100)
print('Accuracy when tf-idf vectorizer paramater set to stopwords(English) using SVM Classifier: ' ,acc_knn_stopwords / 100)

Accuracy using MultinomialNB Classifier :  0.7738980350504514
Accuracy when tf-idf vectorizer paramater set to bigram using MultinominalNB Classifier:  0.7654
Accuracy when tf-idf vectorizer paramater set to stopwords(English) using MultinominalNB Classifier:  0.8169
  
Accuracy using SVM Classifier :  0.9257
Accuracy when tf-idf vectorizer paramater set to bigram using SVM Classifier  :  0.9209999999999999
Accuracy when tf-idf vectorizer paramater set to stopwords(English) using SVM Classifier:  0.9244


#### Observations :
##### Accuracy of the model is high when we used SVM classifier than MultinomialNB Classifier
   1. Accuracy using MultinomialNB Classifier :  0.77
   2. Accuracy using SVM Classifier :  0.92
   
#####  Accuracy of the model is less when tf-idf parameter set to bigram using Multinominal Classifier than the unigram
   1. Accuracy with parameter set to bigram using MultinominalNB Classifier : 0.76
   2. Accuracy of model without parameter changes using MultinominalNB Classifier : 0.77

##### Accuracy of the model is more when tf-idf parameter set to stopwords using Multinominal Classifier than the bigram
   1. Accuracy with parameter set to stopwords using MultinominalNB Classifier : 0.81
   2. Accuracy with parameter set to bigram using MultinominalNB Classifier : 0.76
  
##### Accuracy of the model is equal when tf-idf parameter set to bigram ,stopwords using SVC Classifier to the normal
   1. Accuracy with parameter set to bigram using SVC Classifier : 0.92
   2. Accuracy with parameter set to stopwords using SVC Classifier : 0.92
   3. Accuracy of model without parameter changes using SVC Classifier : 0.92
   
   
 